sdk/storage/azure-storage-common/inc/azure/storage/common/paged_response.hpp

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <type_traits>
+
+#include <azure/core/context.hpp>
+#include <azure/core/http/raw_response.hpp>
+
+namespace Azure { namespace Storage {
+
+  namespace _internal {
+    template <class Derived> class PageResult {
+    public:
+      std::unique_ptr<Azure::Core::Http::RawResponse> RawResponse;
+      std::string NextPageToken;
+      std::string CurrentPageToken;
+
+      bool HasMore() const { return !NextPageToken.empty(); }
+
+      void NextPage(const Azure::Core::Context& context)
+      {
+        if (!HasMore())
+        {
+          // NextPage() should NEVER be called if HasMore() is false
+          std::abort(); // or throw std::out_of_range
+        }
+
+        static_assert(
+            std::is_base_of<PageResult, Derived>::value,
+            "The template argument \"Derived\" should derive from PageResult<Derived>.");
+
+        CurrentPageToken = NextPageToken;
+        static_cast<Derived*>(this)->OnNextPage(context);
```

Comment:
Consider using `dynamic_cast` instead. It helps code auditing sine we minimize static_cast usage. Could we merge the static_assert above if we use dynamic_cast?

Reply:
If we know the correct type, we should use `static_cast` rather than `dynamic_cast`. `dynamic_cast` has runtime overhead.